lib/timers.js

```diff
@@ -554,67 +557,73 @@ function tryOnImmediate(immediate, queue) {
   }
 }
 
+function runCallback(timer) {
+  var argv = timer._argv;
+  var argc = argv ? argv.length : 0;
+  switch (argc) {
+    // fast-path callbacks with 0-3 arguments
+    case 0:
+      return timer._callback();
+    case 1:
+      return timer._callback(argv[0]);
+    case 2:
+      return timer._callback(argv[0], argv[1]);
+    case 3:
+      return timer._callback(argv[0], argv[1], argv[2]);
+    // more then 3 arguments run slower with .apply
+    default:
+      return timer._callback.apply(timer, argv);
+  }
+}
 
-function Immediate() { }
-
-Immediate.prototype.domain = undefined;
-Immediate.prototype._onImmediate = undefined;
-Immediate.prototype._idleNext = undefined;
-Immediate.prototype._idlePrev = undefined;
 
+function createImmediate(callback, args) {
+  return {
+    _idleNext: null,
+    _idlePrev: null,
+    _callback: callback,
+    _argv: args,
+    _onImmediate: callback,
+    domain: process.domain,
+  };
+}
 
 exports.setImmediate = function(callback, arg1, arg2, arg3) {
   if (typeof callback !== 'function') {
     throw new TypeError('"callback" argument must be a function');
   }
 
   var i, args;
-  var len = arguments.length;
-  var immediate = new Immediate();
-
-  L.init(immediate);
 
-  switch (len) {
+  switch (arguments.length) {
     // fast cases
     case 0:
     case 1:
-      immediate._onImmediate = callback;
       break;
     case 2:
-      immediate._onImmediate = function() {
-        callback.call(immediate, arg1);
-      };
+      args = [arg1];
       break;
     case 3:
-      immediate._onImmediate = function() {
-        callback.call(immediate, arg1, arg2);
-      };
+      args = [arg1, arg2];
       break;
     case 4:
-      immediate._onImmediate = function() {
-        callback.call(immediate, arg1, arg2, arg3);
-      };
+      args = [arg1, arg2, arg3];
       break;
     // slow case
     default:
+      var len = arguments.length;
```

Comment:
@benjamingr I changed the three lines that copy the callback params out of `arguments`,
for an added 2x speedup of the "many" (more than 3) arguments handling.  (And yes, it's
still faster to special-case the 1/2/3 handling, but 4/5 no longer, so 4 is the crossover.)

Turns out node v6 is much much faster to `.apply` args in an array that was dynamically
extended than if it was allocated to the exact size.  Seems to have changed with v6;
v5 and v4 were faster to .apply arrays that were created the correct size.
The change is significant, now 4x slower than before.

From the commit message:
```
        // faster to copy args, but v6 very slow to .apply them:
        var argv = new Array(arguments.length - 1);   // exact size
        for (var i = 1; i < arguments.length; i++)
          argv[i - 1] = arguments[i];
    
        // slower to copy args, but v6 much faster to .apply them:
        var argv = new Array();                       // grow as needed
        for (var i = 1; i < arguments.length; i++)
          argv[i - 1] = arguments[i];
```